Frameworks/CoreGraphics/DWriteWrapper.mm

```diff
@@ -580,29 +631,232 @@ HRESULT STDMETHODCALLTYPE CreateStreamFromKey(_In_ const void* fontFileReference
+                                                      _In_ UINT32 collectionKeySize,
+                                                      _Out_ IDWriteFontFileEnumerator** enumerator) {
+        MakeAndInitialize<DWriteFontFileEnumerator>(enumerator, m_fontDatas.get(), &m_previouslyCreatedFiles);
+        return S_OK;
+    }
+
+private:
+    // Array of CFDataRef containing data of fonts, in order of being added
+    woc::unique_cf<CFMutableArrayRef> m_fontDatas;
+
+    // Set of CFDataRef containing data of fonts, used to simplify checking if font has been added
+    woc::unique_cf<CFMutableSetRef> m_fontDatasSet;
+
+    // Array of previously created font files, which saves us from having to read a file multiple times
+    std::vector<ComPtr<IDWriteFontFile>> m_previouslyCreatedFiles;
+};
+
+// TLambda is a member function of our FontCollectionLoader which updates the internal state of the loader
+template <typename TLambda>
+static HRESULT __DWriteUpdateUserCreatedFontCollection(CFArrayRef datas, CFArrayRef* errors, TLambda&& func) {
+    ComPtr<IDWriteFactory> dwriteFactory;
+    RETURN_IF_FAILED(DWriteCreateFactory(DWRITE_FACTORY_TYPE_SHARED, __uuidof(IDWriteFactory), &dwriteFactory));
+
+    // Create singleton Font Collection loader, which will be shared for registering/unregistering fonts
+    static ComPtr<DWriteFontCollectionLoader> loader;
+    static HRESULT createdLoader = [](DWriteFontCollectionLoader** collectionLoader) {
+        ComPtr<IDWriteFactory> dwriteFactory;
+        RETURN_IF_FAILED(DWriteCreateFactory(DWRITE_FACTORY_TYPE_SHARED, __uuidof(IDWriteFactory), &dwriteFactory));
+        RETURN_IF_FAILED(MakeAndInitialize<DWriteFontCollectionLoader>(collectionLoader));
+        RETURN_IF_FAILED(dwriteFactory->RegisterFontCollectionLoader(*collectionLoader));
+        return S_OK;
+    }(&loader);
+    RETURN_IF_FAILED(createdLoader);
+
+    // Call member function to update loader's font data
+    func(loader.Get(), datas, errors);
+
+    // DWrite won't create a new font collection unless we provide a new key each time
+    static size_t collectionKey = 0;
+    RETURN_IF_FAILED(
+        dwriteFactory->CreateCustomFontCollection(loader.Get(), &(++collectionKey), sizeof(collectionKey), &__UserCreatedFontCollection));
+
+    // Update __UserFontPropertiesMap with new values
+    woc::unique_cf<CFArrayRef> userFamilyNames{ __DWriteCopyFontFamilyNamesFromCollection(__UserCreatedFontCollection) };
+    __UserFontPropertiesMap = __CreatePropertiesMapForFamilyNames(userFamilyNames.get());
+
+    return S_OK;
+}
+
+// Registers user defined fonts to a collection so they can be created later
+// Expects datas to be array of CFDataRefs, errors to be out pointer to array of CFErrorRef
+HRESULT _DWriteRegisterFontsWithDatas(CFArrayRef datas, CFArrayRef* errors) {
+    return __DWriteUpdateUserCreatedFontCollection(datas, errors, std::mem_fn(&DWriteFontCollectionLoader::AddDatas));
```

Comment:
I see your point, but it feels silly to create a lambda just to do this, and given it's internal and documented I don't think it will pose an issue #WontFix